crates/wassette/src/component_storage.rs

```diff
@@ -0,0 +1,248 @@
+        let path = self.metadata_path(component_id);
+        if !path.exists() {
+            return Ok(None);
+        }
+
+        let contents = tokio::fs::read_to_string(&path).await.with_context(|| {
+            format!("Failed to read component metadata from {}", path.display())
+        })?;
+
+        let metadata = serde_json::from_str(&contents).with_context(|| {
+            format!("Failed to parse component metadata from {}", path.display())
+        })?;
+        Ok(Some(metadata))
+    }
+
+    /// Write precompiled component bytes to disk.
+    pub async fn write_precompiled(&self, component_id: &str, bytes: &[u8]) -> Result<()> {
+        let path = self.precompiled_path(component_id);
+        tokio::fs::write(&path, bytes).await.with_context(|| {
+            format!(
+                "Failed to write precompiled component to {}",
+                path.display()
+            )
+        })
+    }
+
+    /// Remove a file if it exists, translating IO errors into `anyhow`.
+    pub async fn remove_if_exists(
+        &self,
+        path: &Path,
+        description: &str,
+        component_id: &str,
+    ) -> Result<()> {
+        match tokio::fs::remove_file(path).await {
+            Ok(()) => {
+                tracing::debug!(component_id = %component_id, path = %path.display(), "Removed {}", description);
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                tracing::debug!(component_id = %component_id, path = %path.display(), "{} already absent", description);
+            }
+            Err(e) => {
+                return Err(anyhow!(
+                    "Failed to remove {} at {}: {}",
+                    description,
+                    path.display(),
+                    e
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// Create a validation stamp for a component artifact to track stale data on disk.
+    ///
+    /// When `include_hash` is `true` the SHA-256 hash of the file is
+    /// recorded in addition to size and modification time so changes can be
+    /// detected even when timestamps are unreliable.
+    pub async fn create_validation_stamp(
+        path: &Path,
+        include_hash: bool,
+    ) -> Result<ValidationStamp> {
+        let metadata = tokio::fs::metadata(path)
+            .await
+            .with_context(|| format!("Failed to read metadata for {}", path.display()))?;
+
+        let file_size = metadata.len();
+        let mtime = metadata
+            .modified()
+            .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            .and_then(|t| {
+                t.duration_since(std::time::UNIX_EPOCH)
+                    .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            })?
+            .as_secs();
+
+        let content_hash = if include_hash {
+            let bytes = tokio::fs::read(path)
```

Comment:
Reading the whole file into memory is going to take a bunch of memory. I'd recommend opening the file, then turning it into a `std` file and using `spawn_blocking` to do something like this (haven't checked if it compiles):

```rust
let file = tokio::fs::File::open(path).await?;
let mut file = file.into_std().await?;
let hash = tokio::task::spawn_blocking(move || {
	let mut hasher = sha2::Sha256::new();
	std::io::copy(&mut file, &mut hasher)?;
	return Ok(hasher.finalize())
}).await??;
```

`Sha256` implements `Write` so this should work and be more efficient since it isn't storing anything but the hash in memory